### PR TITLE
fixes for inductor <> batch norm

### PR DESCRIPTION
**Pull request by bdhirsh**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89532
* #89760
* __->__ #89603

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Sounds good!

**Comment by bdhirsh** [comment]
@anijain2305 / @ngimel This PR enables the decomp for batch norm in inductor, and I see at least a few models failing accuracy due to it

I also confirmed locally that `python benchmarks/dynamo/timm_models.py --accuracy --device cuda --backend=inductor --training --float32 --only=resnest101e` passes, but **only** when I modify the benchmark script to use fp64.

(from timm)
dm_nfnet_f0
gmixer_24_224
mixer_b16_224

I talked to animesh about this a bit offline, and one solution is to relax the tolerance a bit on our accuracy tests. What do we think?

Although... looking at the failure for `mixer_b16_224` in CI, I see:
```
Error: ain mixer_b16_224                      [2022-11-24 01:38:05,228] torch._dynamo.utils: [ERROR] RMSE (res-fp64): inf, (ref-fp64): 0.00030 and shape=torch.Size([768, 3, 16, 16])
```

The `inf` seems bad. I think we should get in agreement on whether:

(1) The decomp is ok as is, with its deviations in precision
(2) we should figure out a way to rewrite parts of the decomp to improve its precision, in a way that still makes it amenable to being compiled by inductor.

**Comment by soumith** [comment]
probably (2)

**Comment by anijain2305** [comment]
The nan seems pretty bad. We will have to root cause this.

**Comment by ngimel** [comment]
@bdhirsh are the failures new on this PR? @ezyang said that gmixer_24_224 and mixer_b16_224 are broken on master.

**Comment by ezyang** [comment]
yeah, if it's those two, it's definitely a master problem (though possibly induced by your aot autograd input mutation or batch norm decomp prs which are on master; although i test reverted https://github.com/pytorch/pytorch/pull/89651 and it didn't help.

**Comment by bdhirsh** [comment]
Yep - I'll rebase / check locally to see if any of the new failures are master issues

**Comment by bdhirsh** [comment]
@anijain2305 `dm_nfnet_f0` is the only new model failing accuracy, and the error looks small-ish (?) (0.071 RMS, vs. 0.017 RMS from fp64)

I'll do the following:
(1) confirm locally that `dm_nfnet_f0` passes with fp64
(2) remove `resnest101e` from the list of model failures in CI
(3) go through the other opted out models and quickly check if they now pass with fp64 accuracy

Are we still ok with relaxing accuracy tests a bit? What do you think would be the best way to update the accuracy tests?

**Comment by ezyang** [comment]
Nfnet is known master failure

**Comment by anijain2305** [comment]
Yes, `dm_nfnet` is failing on master. So, please go ahead and merge this when you are ready.

**Comment by bdhirsh** [comment]
cc @msaroufim this PR should fix the test accuracy issues you were seeing in https://github.com/sgugger/torchdynamo-tests/blob/main/scripts/cv_classification.py (would be good to confirm though :) )

It looks like the problem wasn't with the precision of the decomp. It's that we have an existing fix to decompose `batch_norm` -> `native_batch_norm` -> `_native_batch_norm_legit` above autograd, **but** we don't have that fix for when we decompose `batch_norm` -> `cudnn_batch_norm`.

I added a line to decompose `cudnn_batch_norm` above autograd too, and after running the repro, I see test accuracy jump from ~50% to ~95%.

That seems like a pretty good indication that the existing precision issues with the decomp (which cause `resnest101e` to fail accuracy in our benchmark suite) aren't necessarily a hi-pri issue in practice (but we should still fix the decomp).
